### PR TITLE
Limit number of analyzer threads

### PR DESCRIPTION
**Pull request by uklotzde**
- As discussed on [Zulip](https://mixxx.zulipchat.com/#narrow/stream/109695-_support/topic/analysis.2C.20expected.20behavior.20when.20re-processing/near/167928183)
- The code contains detailed comments about why and how

### CODE REVIEW

File: src/library/analysisfeature.cpp [hunk 1]
```diff
@@ -16,13 +16,33 @@
...
+        // Utilize all available CPU cores
+        return idealThreadCount;
+    } else if (idealThreadCount <= 6) {
+        // Reserve 1 core as headroom
+        return idealThreadCount - 1;
+    } else if (idealThreadCount <= 10) {
+        // Reserve 2 cores as headroom
+        return idealThreadCount - 2;
+    } else {
+        // Never utilize more than 8 cores
```
Comment (JosepMaJAZ) [inline src/library/analysisfeature.cpp]: Why do you want to limit it?  
9th gen intel Core i9 has 8 cores with hyperthreading (16 threads) and processors just keep adding more cores ( as one example, my 6th gen i7 "only" has 4 cores/8 threads and the 8th gen has 6 cores/12 threads) . You can add an "idealThreadCount - 3;" in the else.

File: src/library/analysisfeature.cpp [hunk 2]
```diff
@@ -130,9 +150,14 @@ void AnalysisFeature::activate() {
 
 void AnalysisFeature::analyzeTracks(QList<TrackId> trackIds) {
     if (!m_pTrackAnalysisScheduler) {
+        const int numAnalyzerThreads = numberOfAnalyzerThreads();
+        kLogger.info()
+                << "Using"
```
Comment (JosepMaJAZ) [inline src/library/analysisfeature.cpp]: spaces so that the text looks ok

Comment (uklotzde) [reply src/library/analysisfeature.cpp]: Formatted with to auto formatter. Keeping the various string components on separate lines aligns them.

File: src/library/analysisfeature.cpp (hunk 1 above)
Comment (uklotzde) [reply src/library/analysisfeature.cpp]: The single UI thread becomes the bottleneck when adding more and more worker threads. Exceeding a certain number of cores stalls the ui as reports indicate. With 4 physical(!) cores the response is acceptable. This is why I decided to use 8 as the limit. 16 already seem to be too much. It is still an 8x speedup compared to 2.2.

Comment (Be-ing) [reply src/library/analysisfeature.cpp]: > Exceeding a certain number of cores stalls the ui as reports indicate.

Where are those reports? I don't recall anyone reporting this. I can't verify that there is any problem that this PR is fixing. I only have computers with 4 or 2 cores.

Comment (Be-ing) [reply src/library/analysisfeature.cpp]: FWIW, when I am batch analyzing, I don't really care if the UI is perfectly responsive as long as it doesn't completely freeze. I care more than the analysis finishes sooner.

**Comment by uklotzde** [comment]
[Report 1 - Laggy](https://mixxx.zulipchat.com/#narrow/stream/109695-_support/topic/analysis.2C.20expected.20behavior.20when.20re-processing/near/167792509)
[Report 2 - Freezing](https://mixxx.zulipchat.com/#narrow/stream/109695-_support/topic/analysis.2C.20expected.20behavior.20when.20re-processing/near/167926340)

**Comment by Holzhaus** [comment]
What is the status of this?

**Comment by Be-ing** [comment]
I'm still not convinced this is a great idea. @foss did say on Zulip that he thought the lagging GUI was a problem, but did not indicate that intentionally slowing down the analysis would be an acceptable tradeoff.

File: src/library/analysisfeature.cpp (hunk 1 above)
Comment (Be-ing) [reply src/library/analysisfeature.cpp]: I do not understand any rationale for these arbitrary limits. You may be right that the UI thread becomes a bottleneck with increasing numbers of worker threads, but no one has actually experimented with this. AFAIK @JosepMaJAZ is the only contributor with a CPU that has >4 cores.

**Comment by Holzhaus** [comment]
Freezing GUI is unacceptable IMHO. There might be others solution that can be worth investigating, e.g. increasing the priority of the GUI thread, moving database access bottleneck from the GUI thread into a separate thread, spawing a separate decoupled process for analysis, etc.

But in order to get Mixxx 2.3 ASAP, this could be a valid stopgap solution. @Be-ing If you care more about fast analysis than lag-free GUI, what about making this configurable? The current implementation can be used to determine the default value, but users can increase that value if they want to analyze a large number of tracks.

**Comment by Be-ing** [comment]
How about simplifying the formula and getting rid of the arbitrary upper limit?
```
if (idealThreadCount > 2) {
    idealThreadCount--;
}
```

**Comment by Be-ing** [comment]
I tested batch analysis with this branch and master. I have a 4 core CPU with hyperthreading, so this branch runs 7 analysis threads. The Mixxx GUI is a bit more responsive with this branch, but IMO it is a fairly subtle difference. But this is only on Linux, maybe it is worse on macOS.

**Comment by uklotzde** [comment]
The formula was only a best guess. It should reserve some headroom for the UI and the rest of the system. And to prevent the single, coordinating UI thread from becoming overloaded.

**Comment by rrrapha** [comment]
> How about simplifying the formula and getting rid of the arbitrary upper limit?

It could also be something like this:
`idealThreadCount = QThread::idealThreadCount() / 2 + 1;`

**Comment by uklotzde** [comment]
Closing, because it doesn't seem to be desired. I'm tired of keeping track of pending PRs.

**Pull request closed without merging**